### PR TITLE
Add normative statements for DID Document creation algorithm.

### PR DESCRIPTION
**Pull request by msporny**
Many of the [implementations](https://w3c.github.io/did-test-suite/#M23) we have today have guessed on the particular algorithm to create a did:key-based DID Document (and gotten it right for the most part). Now that we're working on a [continuous integration test suite](https://github.com/w3c-ccg/did-key-test-suite) for the did:key method, we need finalized normative statements to ensure multi-implementation interoperability.

This PR adds the normative statements for the DID Document creation algorithm. The algorithm takes options as an input, such as the desired public key format, to modify the final DID Document that is output.

Look at the "Diff" below -- highlighted yellow sections are new content. It's not fully complete yet -- wording could be improved, the JWK transformation section still needs to be written, but I hope to do that (or ideally, others that have implemented that) to do that in another PR.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html" title="Last updated on Jun 16, 2022, 2:04 AM UTC (49be877)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/51/0df342b...49be877.html" title="Last updated on Jun 16, 2022, 2:04 AM UTC (49be877)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+If the byte length of <em>raw public key bytes</em> does not match the expected
+public key length for the associated <em>multicodec value</em>, an
+INVALID_PUBLIC_KEY_LENGTH error MUST be raised.
+            </li>
+            <li>
+Set the <em>verification method</em> `id` value by concatenating
+<em>identifier</em>, a hash character (`#`), and the <em>multicodec value</em>.
+            </li>
+            <li>
+Set the <em>key encoding type</em> value to the <em>options</em> `type` value
```
Comment (OR13) [inline index.html:419]: Am i reading this that `MultiKey` is the default type? It seems like did key would have highest degree of interop with off the shelf tooling if `JsonWebKey` were the default key type... since there are far more libraries that support JWK than that support non standard key types.

File: index.html [hunk 2]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+if it is set, or `Multikey` if it is not set.
+            </li>
+            <li>
+If <em>options</em> `allowExperimentalPublicKeyTypes` is not set to `true` and
```
Comment (OR13) [inline index.html:423]: this reads a bit weird, but you need to `allowExperimentalPublicKeyTypes` in order to present the key with something other than `MultiKey`, `JsonWebKey` or `Ed25519VerificationKey2020`... so for example, to support backwards compatibility with `Ed25519VVerifcaitonKey2018`... I think this option approach makes sense.

File: index.html [hunk 3]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+<em>key encoding type</em> is not `Multikey`, `JsonWebKey2020`, or
+`Ed25519VerificationKey2020`, an INVALID_PUBLIC_KEY_TYPE error MUST be raised.
+            </li>
+            <li>
+Set the <em>verification method</em> `type` to the <em>key encoding type</em>.
+            </li>
+            <li>
+Set the <em>verification method</em> `controller` to <em>identifier</em>.
```
Comment (OR13) [inline index.html:431]: I suggest being more explicit about DID vs DID URL here... seen a lot of first timer errors related to this.

File: index.html [hunk 4]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+
+          <p>
+The following algorithm can be used to encode a multicodec value to a
+JSON Web Key. A <em>multicodec value</em>, <em>raw public key bytes</em>,
+and <em>options</em> are the required inputs. An <em>encoded JWK</em> is
+the output.
+          </p>
+
+          <p class="issue">
+Add the various was of transforming multicodec values and raw public key bytes
```
Comment (OR13) [inline index.html:492]: Its pretty straight forward, do you not support this today?

basically decode the public key as you do for multibase, then expand it if its expandable (secp256k1, secp256r1, 384r1, 521r1...)

encode the expanded values with `kty` `crv` `x` and `y`.... best to test this backwards and forwards from a library that generated a JWK first... instead of from a library that generated multibase first.

File: index.html [hunk 5]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+            </li>
+            <li>
+Return <em>verification method</em>.
+            </li>
+          </ol>
+
+        </section>
+
+        <section class="normative">
+          <h4>Derive Encryption Key Algorithm</h4>
```
Comment (OR13) [inline index.html:581]: its no clear to me which `option` this applies to.

I would expect to be able to do the following:

```
data0 = did.key.create({ ed26619 } )
data1 = did.key.create({ curve25519 } )
data2= did.key.create({ ed26619, curve25519 } )
```

I think this PR does not address the ambiguity of which of these representations you get, unless you are saying derive always happens if its possible... in which case 0 and 2 and the same.

File: index.html [hunk 6]
```diff
@@ -299,7 +656,7 @@ <h3>Deactivate</h3>
       <h2>Test Vectors</h2>
       <!-- No way to do directory listing on github pages short of generating them client side :-(  Not sure if there's a better way to do this -->
       <p>For a full list of test vectors see <a href="https://github.com/w3c-ccg/did-method-key/tree/master/test-vectors">test vectors</a>.</p>
-      
+
       <section class="informative">
         <h3>Ed25519 / X25519</h3>
```
Comment (OR13) [inline index.html:661]: Same note as above: https://github.com/w3c-ccg/did-method-key/pull/51/files#r895671296

(I'm fine tackling this in another PR just trying to gage intent).

File: index.html [hunk 7]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+                    <td>p256-pub - P-256 public key (compressed)</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1201`</td>
+                    <td>?? bytes</td>
+                    <td>p384-pub - P-384 public key (compressed)</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1202`</td>
+                    <td>?? bytes</td>
```
Comment (OR13) [inline index.html:397]:
```suggestion
                    <td> 65 bytes</td>
```

https://stackoverflow.com/questions/50002149/why-p-521-public-key-x-y-some-time-is-65-bytes-some-time-is-66-bytes#:~:text=521%20bits%20is%2065%20full,byte%20is%200x00%20or%200x01).

Comment (OR13) [reply index.html:397]: seems wrong tho... pretty sure that for P-256.

**Review by OR13 — CHANGES_REQUESTED** [review]
Overall, I'm in support of this PR, and many of my comments are exploratory not obstructing.

The ones comment that I believe needs addressing is:

https://github.com/w3c-ccg/did-method-key/pull/51/files#r895690838

If we are changing example JSON, but not explaining how the member who's values is supposed to be generated, it's basically not possible to approve the change, because its not possible to validate that its correct.

AFAIK, this is pretty simple thing to solve, based on implementations I have seen, this would work.

If the type exists in the JSON, add the context.

`Ed25519VerificationKey2020` -> `https://w3id.org/security/suites/ed25519-2020/v1`
`X25519KeyAgreementKey2020` -> `https://w3id.org/security/suites/x25519-2020/v1`
`JsonWebKey*` -> `https://w3id.org/security/suites/jws-2020/v1`
`MultiKey*` -> `https://w3id.org/security/suites/multikey-2021/v1`
```

In other words, you can generate the whole JSON, then add the context as the last step based on the options the caller chose.

File: index.html [hunk 8]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
 associated DID Document:
         </p>
 
         <pre class="example" title="A DID Document derived from a did:key">
           {
             "@context": [
               "https://www.w3.org/ns/did/v1",
-              "https://w3id.org/security/suites/ed25519-2018/v1",
-              "https://w3id.org/security/suites/x25519-2019/v1"
+              "https://w3id.org/security/suites/ed25519-2020/v1",
```
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:246]: Ah, yes, good catch. 

I had it on my list to specify how the `@context` is generated, but forgot to do that in my bleary-eyed state last night. I'll add how `@context` is generated in the next pass to this PR, most likely using the suggestions you mentioned on how to generate `@context`.

What I'm trying to go for w/ `Multikey` and (ideally, `JsonWebKey`) is to remove all of the per-cryptosuite JSON-LD Contexts... so, a did:key DID Document would only need ONE `@context`... "https://www.w3.org/ns/did/v2" (note the "v2" at the end). That context would define `Multikey` and `JsonWebKey` and that's most likely all we'd need for public key extensibility... no more date-versioned cryptosuite contexts.

However, I can't make that change yet because we don't have buy-in from the VC2WG or DIDWG yet... so, this iteration of the algorithm is a bit more funky than I'd like. It's not ideal, I'm just trying to capture what's out there today with what the ideal design is that we're trying to move towards.

**Comment by mprorock** [comment]
> Overall, I'm in support of this PR, and many of my comments are exploratory not obstructing.
> 
> The ones comment that I believe needs addressing is:
> 
> https://github.com/w3c-ccg/did-method-key/pull/51/files#r895690838

Very much inline with Orie on this one.

File: index.html [hunk 9]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+                    <td>ed25519-pub - Ed25519 public key</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1200`</td>
+                    <td>33 bytes</td>
+                    <td>p256-pub - P-256 public key (compressed)</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1201`</td>
+                    <td>?? bytes</td>
```
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:392]: Yeah, when I went hunting for the normative value at IETF, I couldn't find it. Most of the Internet lore that I read suggested using openssl to generate a key and then hand count the raw compressed bytes in the output (what year is this again? How long has this stuff been in production? :P). I'm sure it's buried in some IETF RFC somewhere, but I gave up after looking for around 30 minutes, hoping someone else that had done a p384 implementation would chime in.

**Review by mprorock — CHANGES_REQUESTED** [review]
Let's avoid context changes unless we have clear context generation requirements.

Would also like to see JWKs handled a little better as a first class citizen (but i think this can probably be done in a follow on PR)

File: index.html (hunk 7 above)
Comment (msporny) [reply index.html:397]: Yeah, I found that article too, but couldn't tell if they were talking about a compressed value (most likely) or some internal Go representation (which would surprise me if it was a compressed value). In any case, probably have to use the openssl trick [mentioned here](https://github.com/w3c-ccg/did-method-key/pull/51/files#r895715511).

File: index.html (hunk 1 above)
Comment (msporny) [reply index.html:419]: The implementing software library has the option of picking what the default type is. This line is there to ensure that if the software library doesn't set a default type, that one is asserted. I was trying to avoid the exact issue that you're driving at -- a fight over what the defaults should be. Rather than have that fight, it's up to the software implementer to pick reasonable defaults for their library by defaulting `key encoding type` to whatever they think is best. Transmute will pick JsonWebKey, DB will pick Multikey.

Perhaps I should add some language describing this, or change the algorithm to say that `key encoding type` MUST be set to a default value by the implementation, or an error must be thrown?

File: index.html (hunk 2 above)
Comment (msporny) [reply index.html:423]: Yes, that's the intent of the language... and yes, it reads weird -- I'll try to make another pass at the language to make it easier to read. I was playing around w/ the use of abstract variables, but then had to use concrete ones (like `allowExperimentalPublicKeyTypes`)... so right now, it's a mish-mash of both, which I don't think is helping understandability. I'll try to go back and look at what some of the HTML5 and JSON-LD 1.1 Algorithms did here.

File: index.html (hunk 3 above)
Comment (msporny) [reply index.html:431]: Hrm, good point. What are the first timer errors you're seeing so I can try to craft some language in that direction? 

It feels like this sort of language is way more important for did:web than it is for did:key... but I'm probably missing some nuance in the point you're making.

File: index.html (hunk 4 above)
Comment (msporny) [reply index.html:492]: > Its pretty straight forward, do you not support this today?

No, our libraries do not support JWK key expression, AFAIK.

> basically decode the public key as you do for multibase, then remove EC point compression.
> 
> encode the expanded values with `kty` `crv` `x` and `y`.... best to test this backwards and forwards from a library that generated a JWK first... instead of from a library that generated multibase first.

Sure, but the details matter here (like you have to switch off of the multicodec value, and X25519 differs slightly from other digital signing keys)... so this section has to be very explicit about the translation from multikey to JWK. I didn't want to write something that folks doing JWK stuff would disagree with. Ideally, we'd have someone that has implemented the JWK stuff write this part of the algorithm, which is why I'd rather have it done as another PR.

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:581]: > its not clear to me which option this applies to.

See steps 5 and 9 in Section 3.1.1 Document Creation Algorithm, namely this text:

> 5) Initialize the encryption verification method to the result of passing identifier, multibase value, and options to [§ 3.1.5 Encryption Method Creation Algorithm](https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#encryption-method-creation-algorithm).
> ...
> 9) If options disableKeyAgreement is not set to true, add the encryption verification method value to the verificationMethod array and initialize the keyAgreement property in document to an array where the first item is the value of the id property in encryption verification method.

https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#document-creation-algorithm

... which then involves step 2 in Section 3.1.5 Encryption Method Creation Algorithm, namely this text:

> Set multicodec value and raw public key bytes to the result of passing multibase value and options to [§ 3.1.6 Derive Encryption Key Algorithm](https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#derive-encryption-key-algorithm).

So, at present, derive always happens, but keyAgreement isn't set if `disableKeyAgreement` is set to true.

We should probably change this such that key derivation doesn't happen if `disableKeyAgreement` is set to true.

I'll try to rework that language to be a bit more clear about setting the keyAgreement value. Would that work for you? What would you prefer?

File: index.html (hunk 6 above)
Comment (msporny) [reply index.html:661]: Yeah, it's a good question, let's try to tackle answering that question in this PR.

I'm trying to be sensitive to the fact that IETF is saying that key derivation is fine (via the OSCORE work) and they're providing proofs (THORMAKER) to that effect, while FIPS is (frustratingly) not going to say that's ok for another decade or more. 

So, the best I think we can do is provide a flag that allows implementations to disable key derivation for Ed25519 -> X25519 in order to comply to FIPS audits.

**Comment by msporny** [comment]
@OR13 wrote:
> Overall, I'm in support of this PR, and many of my comments are exploratory not obstructing.

Thanks for the thorough review. I've responded to all of your comments and will make changes in my next pass on this PR.

@mprorock wrote:
> Let's avoid context changes unless we have clear context generation requirements.

Yep, will provide clear generation requirements in a revision to this PR as noted [here](https://github.com/w3c-ccg/did-method-key/pull/51#discussion_r895712637).

> Would also like to see JWKs handled a little better as a first class citizen (but i think this can probably be done in a follow on PR)

Yep, responded to that comment [here]().

File: index.html (hunk 1 above)
Comment (OR13) [reply index.html:419]: I think this is a fight that needs to be resolved to protect the method. 
I feel strongly that a default that is not well supported will harm the method.
I like the language asserting that a default must exist, and that it should be `MultiKey`... except that it says it should be `MultiKey`.

Comment (OR13) [reply index.html:419]: If you like, we can move it to an issue, like you did for JWK production rules.

File: index.html (hunk 2 above)
Comment (OR13) [reply index.html:423]: I got it, just had to read it a few times, you can resolve this comment.

File: index.html (hunk 4 above)
Comment (OR13) [reply index.html:492]: We can leave it as an issue, especially if your implementations don't support this, I expect we or other JWK users will need to define this.

File: index.html (hunk 5 above)
Comment (OR13) [reply index.html:581]: yeah, lets split this up, there are 2 things happening (our libraries already allow you to ask for variations of these for NIST curves / X25519).

1. Do you want key agreement?
2. Do you want key derivation?

If you ask for key agreement but don't allow for derivation for ed25519 you should get an error.

If you ask for derivation with P256, you should get an error.

Point conversion should not be a side effect of asking for key agreement, it should be an explicit request, or an explicit default.

File: index.html (hunk 6 above)
Comment (OR13) [reply index.html:661]: yeah agree, the citations help, let's just make sure the normative requirements are explicit enough to support either posture.

File: index.html (hunk 7 above)
Comment (OR13) [reply index.html:397]: cc @clehner @troyronda halp!

File: index.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:392]: yeah, not making IETF look that great, when it doesn't show up as the top result... the fact that this is hard to answer is a failure.

File: index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:246]: We can defer with an issue, like you did elsewhere for JWK.

**Comment by OR13** [comment]
Reviewed responses, waiting on new commits to see feedback (either adding issues inline, or implementing feedback inline).

File: index.html (hunk 3 above)
Comment (msporny) [reply index.html:431]: Got it, I'll put in language to make those MUSTs explicit.

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:581]: > Point conversion should not be a side effect of asking for key agreement, it should be an explicit request, or an explicit default.

But isn't the only way you get key agreement w/ did:key via key derivation? Why would you pick `allowKeyDerivation=true` but then do `addKeyAgreement=true`?

File: index.html (hunk 1 above)
Comment (msporny) [reply index.html]: > I think this is a fight that needs to be resolved to protect the method.

Ugh, no thanks. :)

I've removed the `Multikey` default and suggested that implementations should pick a reasonable default. If they don't pick a reasonable default, the algorithm throws an error now:

Addressed in 1250253de7506288bc645ff8c2a8a693461369e0

File: index.html (hunk 2 above)
Comment (msporny) [reply index.html]: Added some language in 2bcf5d7d2a50f92396050b7a514dd97f8d67d9a6 in an attempt to make it easier to read. Still don't think it helped much. 

I also tried moving to easier to read variable names using dot-notation like the Fetch spec does and I think that helped a bit. Those changes are in 63f46992c51bbb2de022e720f6f82632bcb831df.

File: index.html (hunk 3 above)
Comment (msporny) [reply index.html]: Fixed in a6aa2cd643afd1e46a865f648ceee0de9ef973e9.

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:609]: Took a cut at it here by adding a new `enableEncryptionKeyDerivation` flag and switching whether `keyAgreement` is set off of that: cc09b8eb45f97192fe79dc3bc44bbbbfed592fe1

File: index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:246]: I went ahead and defined enough to make the example in the spec valid now: 2d2e1016b2aa90094cd68a6c30fc166c2b0f8de1

I couldn't make Multikey work because it's not defined in any context yet, but noted that we expect it to be defined by VC2WG in the next year or so.

**Review by mprorock — COMMENTED** [review]
I have to read this through in a bit more detail, but this looks much closer! thanks @msporny

File: index.html (hunk 5 above)
Comment (OR13) [reply index.html:616]: > But isn't the only way you get key agreement w/ did:key via key derivation?

No, NIST Curves support keyAgreement with the same key they use to support signatures.

Comment (OR13) [reply index.html:616]: `enableEncryptionKeyDerivation` I suppose you could use this to turn keyAgreement on or off for P-256?

File: index.html (hunk 6 above)
Comment (OR13) [reply index.html:756]: Resolving to focus on remaining comments.

**Review by OR13 — APPROVED** [review]

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:616]: > No, NIST Curves support keyAgreement with the same key they use to support signatures.

Sure, but that would never pass a FIPS audit -- it violates the "don't use the same key for signing and KEM" guidance -- and it's especially bad for P-256, that knowingly weakens the shared key material.

> enableEncryptionKeyDerivation I suppose you could use this to turn keyAgreement on or off for P-256?

Yeah, but I guess that would probably be a misuse of that particular flag.

Perhaps we need an `onlyGenerateKeyAgreement` flag for P-256, so you can generate a did:key document for signing OR you can generate a did:key document only for `keyAgreement`, but you should definitely not generate one for both.

Another possibility is to create a variation of did:key that encodes the signing public key and the encryption key in the did:key itself, perhaps separated by a colon, so a P-256 did:key that supports both signing and encryption could look like this:

`did:key:zDnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169:zDnaerx9CtbPJ1q36T5Ln5wYt3MQYeGRG5ehnPAmxcf5mDZpv`

... but that's sorta gross.

We should probably raise an issue for this. I didn't think anyone was trying to support both signing and encryption with did:keys and any non-Montgomery coordinate system.

Comment (mprorock) [reply index.html:616]: We should explicitly disallow signing and encryption from a single did:key

**Review by mprorock — APPROVED** [review]
@msporny want a separate PR around key material handling?

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:616]: > We should explicitly disallow signing and encryption from a single did:key

Except where there are IETF documents by CFRG folks with mathematical proofs showing that it's safe to do so, such as with ECDH w/ Montgomery Curve public key derivation (OSCORE, THORMARKER), but with an option to turn that feature off for those that need to demonstrate FIPS compliance, right? :)

File: index.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (clehner) [reply index.html]: Length and serialization in general is in here: http://www.secg.org/download/aid-780/sec1-v2.pdf *Standards for Efficient Cryptography 1 (SEC 1) - 2.3.3. Elliptic-Curve-Point-to-Octet-String Conversion*

> mlen = ⌈(log2 q)/8⌉+1 if P ̸= O and point compression is used

and parameters for P-384 are here (D.1.2.4 Curve P-384):
https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf

... although I admit I don't fully understand how these fit together; but if "q" is 384 bits in length, then mlen is 49 (1 byte for the format tag / compressed y coordinate, followed by 48 bytes (384 bits) for the x coordinate).

Suggesting 49 bytes in https://github.com/w3c-ccg/did-method-key/pull/51/files#r896817820.

File: index.html [hunk 10]
```diff
@@ -232,39 +235,488 @@ <h3>Create</h3>
...
-                "publicKeyBase58": "C2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w"
+                "publicKeyMultibase": "zC2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w"
               }
             ],
             "assertionMethod": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2"],
             "authentication": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2"],
             "capabilityInvocation": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2"],
             "capabilityDelegation": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2"],
-            "keyAgreement": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6LSni4ALZW58RuBPizpNeTK3HGfZfSuE3zcxeBubnrssYuh"]
+            "keyAgreement": ["did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#zC2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w"]
```
Comment (dlongley) [inline index.html:268]: Same issue here -- the hash fragment should start with `z6LS`.

**Review by dlongley — APPROVED** [review]
Thanks for doing this! We'll need to confirm the steps in our own implementation to see if there are any differences.

File: index.html [hunk 11]
```diff
@@ -232,39 +235,488 @@ <h3>Create</h3>
...
+                    <td>ed25519-pub - Ed25519 public key</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1200`</td>
+                    <td>33 bytes</td>
+                    <td>p256-pub - P-256 public key (compressed)</td>
+                  </tr>
+                  <tr>
+                    <td>`0x1201`</td>
+                    <td>?? bytes</td>
```
[the remark this answers is not in the document]
Comment (clehner) [reply index.html]: As described in https://github.com/w3c-ccg/did-method-key/pull/51/files#r896860648.

---

Implementation: https://github.com/spruceid/ssi/blob/main/src/jwk.rs#L965
https://github.com/spruceid/ssi/blob/acd3f21f44a596498cb19e0dd5e1d5c5c43255cd/did-key/src/lib.rs#L173

---

Checking test vector: https://w3c-ccg.github.io/did-method-key/#p-384
https://github.com/w3c-ccg/did-method-key/blob/0df342b31f0557eeb6c9fe2b780b5f19bdf9ae5d/index.html#L352-L357

```
$ echo -n 82Lm1MpAkeJcix9K8TMiLd5NMAhnwkjjCBeWHXyu3U4oT2MVJJKXkcVBgjGhnLBn2Kaau9|base58 -d | wc -c
51
$ echo -n 82Lm1MpAkeJcix9K8TMiLd5NMAhnwkjjCBeWHXyu3U4oT2MVJJKXkcVBgjGhnLBn2Kaau9|base58 -d | hexdump -C
00000000  81 24 03 94 89 d3 c6 5f  1f 8c b2 a9 fd 40 ab c4  |.$....._.....@..|
00000010  8d 16 0e 49 5a 86 2f bb  fb fe 89 6e d8 87 8e 24  |...IZ./....n...$|
00000020  6f 32 f8 5e 76 12 95 74  70 5f 8b 61 c2 63 cb 7f  |o2.^v..tp_.a.c..|
00000030  40 9c 9c                                          |@..|
00000033
```
51 bytes: Two bytes (`0x8124`) for multicodec varint tag (`0x1201`), followed by one byte compressed y-coordinate / format tag (`0x03`), followed by 48 bytes of key material (x-coordinate).

---

Compare to P-256: https://w3c-ccg.github.io/did-method-key/#p-256
https://github.com/w3c-ccg/did-method-key/blob/0df342b31f0557eeb6c9fe2b780b5f19bdf9ae5d/index.html#L343-L348
```
$ echo -n DnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169|base58 -d | wc -c
35
$ echo -n DnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169|base58 -d | hexdump -C
00000000  80 24 03 7f 23 58 30 dd  3d ef a7 22 ef 1a a2 49  |.$..#X0.=.."...I|
00000010  d6 a0 dd bb a4 f9 90 b0  81 75 38 93 3f 57 36 40  |.........u8.?W6@|
00000020  65 35 42                                          |e5B|
00000023
```
35 bytes: `0x8024` (multicodec varint encoding of `0x1200`), `0x03` tag (compressed y-coordinate) and 32 bytes x-coordinate.

File: index.html (hunk 5 above)
Comment (OR13) [reply index.html:616]: I'm not sure point conversion will pass a FIPS audit either, best not to require it or default to it.... instead allow for it to be configured.

If you want to use the same did:key for signing and encryption you can. 

If you want to get a presentation of a did key that only supports signing (or only supports encryption) you should be able to do this... BTW, a better solution is to just use the x2559 key as its own separate did... like this:

https://did.key.transmute.industries/generate/x25519?seed=d197db5d456b1be622144c8f0b6b12602fe93b29e65d8e47a15d67abe8513fbc

https://did.key.transmute.industries/generate/ed25519?seed=49f85b0fdce80cfed8db5daad14e42d4d9493add3b95281dc56a60ebd2659d5d

This allows you to: use signing and encryption with did key, not reuse entropy, not do point conversion... this is likely to be looked on more favorably during an audit.

Comment (OR13) [reply index.html:616]: todays's did key implementations don't allow you to isolate the key purpose, and the ed25519 variant actually encourages entropy reuse under the assumption that point conversion is safe... this is essentially treating experimental behavior as a safe default... and still violating the "use a key for a single purpose" NIST recommendation.

Comment (OR13) [reply index.html:616]: The nature of the did key spec is that its an identifer for a public key... not an identifer for nist recommendations of a public key and its use... better to use did:jwk for that, and set the `use` property before encoding... https://github.com/quartzjer/did-jwk/blob/main/spec.md

File: index.html [hunk 12]
```diff
@@ -232,39 +235,488 @@ <h3>Create</h3>
...
+                "publicKeyMultibase": "z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2"
               },
               {
-                "id": "did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#z6LSni4ALZW58RuBPizpNeTK3HGfZfSuE3zcxeBubnrssYuh",
-                "type": "X25519KeyAgreementKey2019",
+                "id": "did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2#zC2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w",
+                "type": "X25519KeyAgreementKey2020",
                 "controller": "did:key:z6MkicdicToW5HbxPP7zZV1H7RHvXgRMhoujWAF2n5WQkdd2",
-                "publicKeyBase58": "C2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w"
+                "publicKeyMultibase": "zC2szpFhD2yBSJLd3qzwMih4BiWunXSpU5fUE7LDMAB8w"
```
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:261]: agreed, example (note the identifier, not the base58 encoded part).

https://did.key.transmute.industries/generate/x25519?seed=4223e087f379609ca0a46dddbb599eab5b3553fc58c5b0ef3780a6b64475535f

File: index.html (hunk 5 above)
Comment (dlongley) [reply index.html:616]: @OR13,

> The ed25519 variant actually encourages entropy reuse under the assumption that point conversion is safe...

I recommend that we start saying "...based on a [proof](https://eprint.iacr.org/2021/509) that point conversion is safe..." instead of "under the assumption" to avoid misleading people that there is no such proof and it is a bare assumption only.

Comment (dlongley) [reply index.html:616]: I should also add -- expressing the same key in multiple ways is not a violation -- and that's all that we're talking about here.  Whether or not that key is *used* is a separate matter in my view -- and unrelated to what this spec is doing. This spec does not say "you MUST use this key" nor would that be enforceable.

File: index.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (dlongley) [reply index.html:261]: Another example of a `did:key` DID doc with x25519 key agreement key:

https://github.com/digitalbazaar/did-method-key#example-did-document

[the remark this answers is not in the document]
Comment (OR13) [reply index.html:261]: ^ that example uses the same entropy for both signing and encryption keys, via point conversion correct? 

Do you have any examples of NIST curves?

File: index.html [hunk 13]
```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
...
+            <li>
+Initialize <em>verification method</em> to an empty object.
+            </li>
+            <li>
+Set <em>multicodec value</em> and <em>raw public key bytes</em> to the result
+of passing <em>multibase value</em> and <em>options</em> to
+<a href="#decode-public-key-algorithm"></a>.
+            </li>
+            <li>
+Ensure the proper key length of <em>raw public key bytes</em> based on the
```
Comment (kdenhartog) [inline index.html]: It's safer to validate the point actually exists on the curve rather than assuming that the correct byte length is all that's necessary to validate the public key. I'd suggest utilzing a point validation check here rather than a byte length check.

Comment (msporny) [reply index.html]: Yes, agreed @kdenhartog. My only concern is how easy it is to do this check across all public key types that are supported -- for example, does libopenssl key import do the check for all these types of public keys: RSA, Ed25519, X25519, secpr1, secpk1... what does this look like for BLS12381, etc?

I guess the test suite could test with known bad public key values for each public key type... I'm going to try to play with some text to add in a "correctness" test and see where I get to.

File: index.html (hunk 10 above)
Comment (msporny) [reply index.html]: Fixed in 51082d03db60678a0c7d84dfc1cbdfb5fb9b83c6.

File: index.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (msporny) [reply index.html]: Fixed in 51082d03db60678a0c7d84dfc1cbdfb5fb9b83c6.

File: index.html (hunk 13 above)
Comment (msporny) [reply index.html]: Implemented in c981fab6e5769d23c8f936b641bd89725b4bcf53.

**Comment by msporny** [comment]
Multiple positive reviews, comments made and suggestions applied, no objections, merging.

File: index.html (hunk 13 above)
Comment (kdenhartog) [reply index.html]: That makes sense to me. As for point validation I know elliptic does it (because I caught an invalid curve attack in their ECDH implementation because they didn't perform the check even though they implemented it) and I just checked Openssl [has it](https://github.com/openssl/openssl/blob/master/crypto/ec/ec_key.c#L472). The different kinds of checks are listed in [Section 5.6.2.3 of NIST SP 800-56A Rev. 3](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/nist.sp.800-56Ar3.pdf)

File: index.html (hunk 1 above)
Comment (bshambaugh) [reply index.html]: > The implementing software library has the option of picking what the default type is. This line is there to ensure that if the software library doesn't set a default type, that one is asserted. I was trying to avoid the exact issue that you're driving at -- a fight over what the defaults should be. Rather than have that fight, it's up to the software implementer to pick reasonable defaults for their library by defaulting `key encoding type` to whatever they think is best. Transmute will pick JsonWebKey, DB will pick Multikey.
> 
https://www.npmjs.com/package/key-did-resolver uses JsonWebKey2020
e.g. https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L27-L52
